## Import Necessary Libraries

In [1]:
# Import necessary libraries  
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
from scipy.stats import chi2_contingency
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, precision_recall_fscore_support
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
import os
import time


In [2]:
print('program is running')
print()
start_time = time.time()

program is running



##### Understanding the problem statement is the first and foremost step. This would help you give an intuition of what you will face ahead of time. Let us see the problem statement.

## Load the dataset

In [3]:
# Load the dataset
a1 = pd.read_excel("case_study1.xlsx")
a2 = pd.read_excel("case_study2.xlsx")


## Description of datasets

In [4]:
df1 = a1.copy()
df2 = a2.copy()

In [5]:
df1

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
0               1         5              4              1   
1               2         1              0              1   
2               3         8              0              8   
3               4         1              0              1   
4               5         3              2              1   
...           ...       ...            ...            ...   
51331       51332         3              0              3   
51332       51333         4              2              2   
51333       51334         2              1              1   
51334       51335         2              1              1   
51335       51336         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
0                        0                  0            0.000   
1                        0                  0            0.000   
2                        1                  0            0.125   
3                        1                  0            1.000   
4                        0                  0            0.000   
...                    ...                ...              ...   
51331                    1                  0            0.333   
51332                    0                  1            0.000   
51333                    1                  1            0.500   
51334                    0                  0            0.000   
51335                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
0                   0.00          0.200          0.800  ...      0   
1                   0.00          1.000          0.000  ...      0   
2                   0.00          1.000          0.000  ...      0   
3                   0.00          1.000          0.000  ...      0   
4                   0.00          0.333          0.667  ...      0   
...                  ...            ...            ...  ...    ...   
51331               0.00          1.000          0.000  ...      0   
51332               0.25          0.500          0.500  ...      0   
51333               0.50          0.500          0.500  ...      0   
51334               0.00          0.500          0.500  ...      0   
51335               0.00          1.000          0.000  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
0                0        1        0      4           1             4   
1                1        0        0      0           0             1   
2                6        1        0      0           2             6   
3                0        0        0      0           0             1   
4                0        0        0      0           3             0   
...            ...      ...      ...    ...         ...           ...   
51331            2        0        0      0           0             3   
51332            2        0        0      0           2             2   
51333            2        0        0      0           0             2   
51334            2        0        0      0           0             2   
51335            0        0        0      0           1             0   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
0             0             72             18  
1             0              7              7  
2             0             47              2  
3             1              5              5  
4             2            131             32  
...         ...            ...            ...  
51331         1             24              5  
51332         0             74              7  
51333         0              9              5  
51334         0             15              8  
51335         0             20             20  

[51336 rows x 26 columns]

In [6]:
df1.shape

(51336, 26)

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   PROSPECTID            51336 non-null  int64  
 1   Total_TL              51336 non-null  int64  
 2   Tot_Closed_TL         51336 non-null  int64  
 3   Tot_Active_TL         51336 non-null  int64  
 4   Total_TL_opened_L6M   51336 non-null  int64  
 5   Tot_TL_closed_L6M     51336 non-null  int64  
 6   pct_tl_open_L6M       51336 non-null  float64
 7   pct_tl_closed_L6M     51336 non-null  float64
 8   pct_active_tl         51336 non-null  float64
 9   pct_closed_tl         51336 non-null  float64
 10  Total_TL_opened_L12M  51336 non-null  int64  
 11  Tot_TL_closed_L12M    51336 non-null  int64  
 12  pct_tl_open_L12M      51336 non-null  float64
 13  pct_tl_closed_L12M    51336 non-null  float64
 14  Tot_Missed_Pmnt       51336 non-null  int64  
 15  Auto_TL            

In [8]:
df2

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
0               1                        549                           35   
1               2                         47                       -99999   
2               3                        302                           11   
3               4                     -99999                       -99999   
4               5                        583                       -99999   
...           ...                        ...                          ...   
51331       51332                         15                           24   
51332       51333                         57                       -99999   
51333       51334                         32                       -99999   
51334       51335                         58                       -99999   
51335       51336                         74                       -99999   

       time_since_recent_deliquency  num_times_delinquent  \
0                                15                    11   
1                            -99999                     0   
2                                 3                     9   
3                            -99999                     0   
4                            -99999                     0   
...                             ...                   ...   
51331                            23                     2   
51332                        -99999                     0   
51333                        -99999                     0   
51334                        -99999                     0   
51335                        -99999                     0   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
0                         29                         29               0   
1                     -99999                          0               0   
2                         25                         25               1   
3                     -99999                          0               0   
4                     -99999                          0               0   
...                      ...                        ...             ...   
51331                     24                         24               0   
51332                 -99999                          0               0   
51333                 -99999                          0               0   
51334                 -99999                          0               0   
51335                 -99999                          0               0   

       num_deliq_12mts  num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  \
0                    0                  0  ...                     0.0   
1                    0                  0  ...                     0.0   
2                    9                  8  ...                     0.0   
3                    0                  0  ...                     0.0   
4                    0                  0  ...                     0.0   
...                ...                ...  ...                     ...   
51331                0                  0  ...                     0.0   
51332                0                  0  ...                     0.0   
51333                0                  0  ...                     0.0   
51334                0                  0  ...                     0.0   
51335                0                  0  ...                     0.0   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
0                         0.0                     0.0   
1                         0.0                     0.0   
2                         0.0                     0.0   
3                         0.0                     0.0   
4                         0.0                     0.0   
...                       ...                     ...   
51331                     0.0                     0.0   
51332                     0.0                     0.0   
51333                     1.0                 

In [9]:
df2.shape

(51336, 62)

In [10]:
df2.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 62 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   PROSPECTID                    51336 non-null  int64  
 1   time_since_recent_payment     51336 non-null  int64  
 2   time_since_first_deliquency   51336 non-null  int64  
 3   time_since_recent_deliquency  51336 non-null  int64  
 4   num_times_delinquent          51336 non-null  int64  
 5   max_delinquency_level         51336 non-null  int64  
 6   max_recent_level_of_deliq     51336 non-null  int64  
 7   num_deliq_6mts                51336 non-null  int64  
 8   num_deliq_12mts               51336 non-null  int64  
 9   num_deliq_6_12mts             51336 non-null  int64  
 10  max_deliq_6mts                51336 non-null  int64  
 11  max_deliq_12mts               51336 non-null  int64  
 12  num_times_30p_dpd             51336 non-null  int64  
 13  n

## Shapes of dataset

In [11]:
print('shapes of data:',df1.shape,df2.shape)

shapes of data: (51336, 26) (51336, 62)


## Checking for null values

In [12]:
df1.isnull().sum()

PROSPECTID              0
Total_TL                0
Tot_Closed_TL           0
Tot_Active_TL           0
Total_TL_opened_L6M     0
Tot_TL_closed_L6M       0
pct_tl_open_L6M         0
pct_tl_closed_L6M       0
pct_active_tl           0
pct_closed_tl           0
Total_TL_opened_L12M    0
Tot_TL_closed_L12M      0
pct_tl_open_L12M        0
pct_tl_closed_L12M      0
Tot_Missed_Pmnt         0
Auto_TL                 0
CC_TL                   0
Consumer_TL             0
Gold_TL                 0
Home_TL                 0
PL_TL                   0
Secured_TL              0
Unsecured_TL            0
Other_TL                0
Age_Oldest_TL           0
Age_Newest_TL           0
dtype: int64

In [13]:
df1.isnull().sum().sum()

0

In [14]:
df2.isnull().sum()

PROSPECTID                      0
time_since_recent_payment       0
time_since_first_deliquency     0
time_since_recent_deliquency    0
num_times_delinquent            0
                               ..
GL_Flag                         0
last_prod_enq2                  0
first_prod_enq2                 0
Credit_Score                    0
Approved_Flag                   0
Length: 62, dtype: int64

In [15]:
df2.isnull().sum().sum()

0

### No null values in both data sets

#### but  we can see -99999 is un wanted value its definetly error in the dataset

In [16]:
df1 = df1.loc[df1['Age_Oldest_TL'] != -99999]

##### removing values with -99999 in the df1 dataset

In [17]:
df1.shape # 10 rows are removed those are removed for the better usage of datasets

(51296, 26)

In [18]:
print(df1.dtypes ,df2.dtypes)

PROSPECTID                int64
Total_TL                  int64
Tot_Closed_TL             int64
Tot_Active_TL             int64
Total_TL_opened_L6M       int64
Tot_TL_closed_L6M         int64
pct_tl_open_L6M         float64
pct_tl_closed_L6M       float64
pct_active_tl           float64
pct_closed_tl           float64
Total_TL_opened_L12M      int64
Tot_TL_closed_L12M        int64
pct_tl_open_L12M        float64
pct_tl_closed_L12M      float64
Tot_Missed_Pmnt           int64
Auto_TL                   int64
CC_TL                     int64
Consumer_TL               int64
Gold_TL                   int64
Home_TL                   int64
PL_TL                     int64
Secured_TL                int64
Unsecured_TL              int64
Other_TL                  int64
Age_Oldest_TL             int64
Age_Newest_TL             int64
dtype: object PROSPECTID                       int64
time_since_recent_payment        int64
time_since_first_deliquency      int64
time_since_recent_deliquency     int6

#### Removing -99999 values in df2 dataset 

In [19]:
columns_to_be_removed = []

for i in df2.columns:
    if df2.loc[df2[i] == -99999].shape[0] > 10000:
        columns_to_be_removed .append(i)


In [20]:
df2 = df2.drop(columns_to_be_removed, axis =1)

##### here drop means delete , axis = 1 means column . 1 = columns, 0 = Rows

In [21]:
df2.shape # 8 columns are deleted

(51336, 54)

In [22]:
for i in df2.columns:
    df2 = df2.loc[df2[i] != -99999]

In [23]:
df2.shape # many rows are deleted because they are null values they ae not much useful in analysis

(42066, 54)

In [24]:
# Checking common column names
for i in list(df1.columns):
    if i in list(df2.columns):
        print (i)

PROSPECTID


### Merge of both datasets to analysis

In [25]:
# Merge the two dataframes, inner join so that no nulls are present
df = pd. merge ( df1, df2, how ='inner', left_on = ['PROSPECTID'], right_on = ['PROSPECTID'] )

In [26]:
df.shape # checking shape of data after merging

(42064, 79)

In [27]:
df

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
0               1         5              4              1   
1               2         1              0              1   
2               3         8              0              8   
3               5         3              2              1   
4               6         6              5              1   
...           ...       ...            ...            ...   
42059       51332         3              0              3   
42060       51333         4              2              2   
42061       51334         2              1              1   
42062       51335         2              1              1   
42063       51336         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
0                        0                  0            0.000   
1                        0                  0            0.000   
2                        1                  0            0.125   
3                        0                  0            0.000   
4                        0                  0            0.000   
...                    ...                ...              ...   
42059                    1                  0            0.333   
42060                    0                  1            0.000   
42061                    1                  1            0.500   
42062                    0                  0            0.000   
42063                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  \
0                   0.00          0.200          0.800  ...   
1                   0.00          1.000          0.000  ...   
2                   0.00          1.000          0.000  ...   
3                   0.00          0.333          0.667  ...   
4                   0.00          0.167          0.833  ...   
...                  ...            ...            ...  ...   
42059               0.00          1.000          0.000  ...   
42060               0.25          0.500          0.500  ...   
42061               0.50          0.500          0.500  ...   
42062               0.00          0.500          0.500  ...   
42063               0.00          1.000          0.000  ...   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0                         0.0                     0.0                   0.000   
1                         0.0                     0.0                   0.000   
2                         0.0                     0.0                   0.000   
3                         0.0                     0.0                   0.000   
4                         1.0                     0.0                   0.429   
...                       ...                     ...                     ...   
42059                     0.0                     0.0                   0.000   
42060                     0.0                     0.0                   0.000   
42061                     1.0                     0.0                   1.000   
42062                     0.0                     0.0                   0.000   
42063                     0.0                     0.0                   0.000   

       pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  last_prod_enq2  \
0                         0.0        1        0              PL   
1                         0.0        0        0    ConsumerLoan   
2                         0.0        1        0    ConsumerLoan   
3                         0.0        0        0              AL   
4                         0.0        1        0    ConsumerLoan   
...                       ...      ...      ...             ...   
42059                     0.0        0        0    ConsumerLoan   
42060                     0.0        0        0          others   
42061                     0.0        0        0    ConsumerLoan   
42062                     0.0        0        0    ConsumerLoan   
42063         

### Checking for categorical data in dataset  

#### categorical values not numericals

In [28]:
# check how many columns are categorical
for i in df.columns:
    if df[i].dtype == 'object':
        print(i)

MARITALSTATUS
EDUCATION
GENDER
last_prod_enq2
first_prod_enq2
Approved_Flag


### For categorical values we use chi-square test to check the relation between features

#### Here we didn't take Approved_flag feature because its target feature

In [29]:
# Chi-square test
for i in ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']:
    chi2, pval, _, _ = chi2_contingency(pd.crosstab(df[i], df['Approved_Flag']))
    print(i, '---', pval)

MARITALSTATUS --- 3.578180861038862e-233
EDUCATION --- 2.6942265249737532e-30
GENDER --- 1.907936100186563e-05
last_prod_enq2 --- 0.0
first_prod_enq2 --- 7.84997610555419e-287


### All p values are less than 0.05 we will accept all categorical features

#### e-233 means for marital status 3* 10 power -233 same as follows

## Varaiation Inflation Factor

In [30]:
# VIF for numerical columns
numeric_columns = []
for i in df.columns:
    if df[i].dtype != 'object' and i not in ['PROSPECTID','Approved_Flag']:
        numeric_columns.append(i)

##### VIF quantifies multicollinearity in regression analysis; values >5 or 10 suggest high multicollinearity, requiring remedial action like variable removal or regularization.

In [31]:
numeric_columns

['Total_TL',
 'Tot_Closed_TL',
 'Tot_Active_TL',
 'Total_TL_opened_L6M',
 'Tot_TL_closed_L6M',
 'pct_tl_open_L6M',
 'pct_tl_closed_L6M',
 'pct_active_tl',
 'pct_closed_tl',
 'Total_TL_opened_L12M',
 'Tot_TL_closed_L12M',
 'pct_tl_open_L12M',
 'pct_tl_closed_L12M',
 'Tot_Missed_Pmnt',
 'Auto_TL',
 'CC_TL',
 'Consumer_TL',
 'Gold_TL',
 'Home_TL',
 'PL_TL',
 'Secured_TL',
 'Unsecured_TL',
 'Other_TL',
 'Age_Oldest_TL',
 'Age_Newest_TL',
 'time_since_recent_payment',
 'num_times_delinquent',
 'max_recent_level_of_deliq',
 'num_deliq_6mts',
 'num_deliq_12mts',
 'num_deliq_6_12mts',
 'num_times_30p_dpd',
 'num_times_60p_dpd',
 'num_std',
 'num_std_6mts',
 'num_std_12mts',
 'num_sub',
 'num_sub_6mts',
 'num_sub_12mts',
 'num_dbt',
 'num_dbt_6mts',
 'num_dbt_12mts',
 'num_lss',
 'num_lss_6mts',
 'num_lss_12mts',
 'recent_level_of_deliq',
 'tot_enq',
 'CC_enq',
 'CC_enq_L6m',
 'CC_enq_L12m',
 'PL_enq',
 'PL_enq_L6m',
 'PL_enq_L12m',
 'time_since_recent_enq',
 'enq_L12m',
 'enq_L6m',
 'enq_L3m',

## VIF sequentially check

In [32]:
# VIF sequentially check

vif_data = df[numeric_columns]
total_columns = vif_data.shape[1]
columns_to_be_kept = []
column_index = 0

### vif can be check by two methods 1.Parallely 2. sequentailly
Here i do sequentailly

In [33]:
for i in range (0,total_columns):
    
    vif_value = variance_inflation_factor(vif_data, column_index)
    print (column_index,'---',vif_value)  
    if vif_value <= 6:
        columns_to_be_kept.append( numeric_columns[i] )
        column_index = column_index+1
    else:
        vif_data = vif_data.drop([ numeric_columns[i] ] , axis=1)  

0 --- inf
0 --- inf
0 --- 11.320180023967996
0 --- 8.363698035000336
0 --- 6.520647877790928
0 --- 5.149501618212625
1 --- 2.611111040579735
2 --- inf
2 --- 1788.7926256209232
2 --- 8.601028256477228
2 --- 3.8328007921530785
3 --- 6.099653381646731
3 --- 5.581352009642762
4 --- 1.9855843530987785
5 --- inf
5 --- 4.809538302819343
6 --- 23.270628983464636
6 --- 30.595522588100053
6 --- 4.3843464059655854
7 --- 3.064658415523423
8 --- 2.898639771299253
9 --- 4.377876915347322
10 --- 2.2078535836958433
11 --- 4.916914200506864
12 --- 5.214702030064725
13 --- 3.3861625024231476
14 --- 7.840583309478997
14 --- 5.255034641721438
15 --- inf
15 --- 7.380634506427232
15 --- 1.421005001517573
16 --- 8.083255010190316
16 --- 1.6241227524040112
17 --- 7.257811920140003
17 --- 15.59624383268298
17 --- 1.8258570471324318
18 --- 1.5080839450032661
19 --- 2.172088834824578
20 --- 2.623397553527229
21 --- 2.2959970812106176
22 --- 7.360578319196446
22 --- 2.1602387773102554
23 --- 2.8686288267891467
24

In [34]:
vif_data

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
0                0.000               0.00                   0   
1                0.000               0.00                   0   
2                0.125               0.00                   0   
3                0.000               0.00                   0   
4                0.000               0.00                   1   
...                ...                ...                 ...   
42059            0.333               0.00                   0   
42060            0.000               0.25                   1   
42061            0.500               0.50                   1   
42062            0.000               0.00                   1   
42063            0.000               0.00                   0   

       pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
0                   0.000                0      0        0      4           1   
1                   0.000                0      0        0      0           0   
2                   0.000                1      0        0      0           2   
3                   0.000                0      0        0      0           3   
4                   0.167                0      0        0      0           6   
...                   ...              ...    ...      ...    ...         ...   
42059               0.000                0      0        0      0           0   
42060               0.250                0      0        0      0           2   
42061               0.500                0      0        0      0           0   
42062               0.500                0      0        0      0           0   
42063               0.000                0      0        0      0           1   

       Unsecured_TL  ...  enq_L3m  NETMONTHLYINCOME  Time_With_Curr_Empr  \
0                 4  ...        0             51000                  114   
1                 1  ...        0             19000                   50   
2                 6  ...        0                18                  191   
3                 0  ...        0             15000                   75   
4                 0  ...        4                 0                  154   
...             ...  ...      ...               ...                  ...   
42059             3  ...        1             18500                  249   
42060             2  ...        0             25000                  186   
42061             2  ...        2             18000                   66   
42062             2  ...        0             12802                   54   
42063             0  ...        1             16000                  102   

       pct_currentBal_all_TL  CC_Flag  PL_Flag  pct_PL_enq_L6m_of_ever  \
0                      0.798        0        1                   0.000   
1                      0.370        0        0                   0.000   
2                      0.585        0        0                   0.000   
3                      0.000        0        0                   0.000   
4                      0.000        0        0                   0.429   
...                      ...      ...      ...                     ...   
42059                  0.230        0        0                   0.000   
42060                  0.660        0        0                   0.000   
42061                  0.428        0        0                   1.000   
42062                  0.143        0        0                   0.000   
42063                  0.333        0        0                   0.000   

       pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  
0                         0.0        1        0  
1                         0.0        0        0  
2                         0.0        1        0  
3                         0.0        0        0  
4                         0.0        1        0  
...                       ...      ...      ...  
42059                     0.0        0        0  
42060                     0.0        0        0  
4

In [35]:
vif_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 39 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            42064 non-null  float64
 1   pct_tl_closed_L6M          42064 non-null  float64
 2   Tot_TL_closed_L12M         42064 non-null  int64  
 3   pct_tl_closed_L12M         42064 non-null  float64
 4   Tot_Missed_Pmnt            42064 non-null  int64  
 5   CC_TL                      42064 non-null  int64  
 6   Home_TL                    42064 non-null  int64  
 7   PL_TL                      42064 non-null  int64  
 8   Secured_TL                 42064 non-null  int64  
 9   Unsecured_TL               42064 non-null  int64  
 10  Other_TL                   42064 non-null  int64  
 11  Age_Oldest_TL              42064 non-null  int64  
 12  Age_Newest_TL              42064 non-null  int64  
 13  time_since_recent_payment  42064 non-null  int

### Anova Test

In [36]:
# check Anova for columns_to_be_kept 
from scipy.stats import f_oneway
columns_to_be_kept_numerical = []

In [37]:
for i in columns_to_be_kept:
    a = list(df[i])  
    b = list(df['Approved_Flag'])  
    
    group_P1 = [value for value, group in zip(a, b) if group == 'P1']
    group_P2 = [value for value, group in zip(a, b) if group == 'P2']
    group_P3 = [value for value, group in zip(a, b) if group == 'P3']
    group_P4 = [value for value, group in zip(a, b) if group == 'P4']


    f_statistic, p_value = f_oneway(group_P1, group_P2, group_P3, group_P4)

    if p_value <= 0.05:
        columns_to_be_kept_numerical.append(i)


In [38]:
columns_to_be_kept_numerical

['pct_tl_open_L6M',
 'pct_tl_closed_L6M',
 'Tot_TL_closed_L12M',
 'pct_tl_closed_L12M',
 'Tot_Missed_Pmnt',
 'CC_TL',
 'Home_TL',
 'PL_TL',
 'Secured_TL',
 'Unsecured_TL',
 'Other_TL',
 'Age_Oldest_TL',
 'Age_Newest_TL',
 'time_since_recent_payment',
 'max_recent_level_of_deliq',
 'num_deliq_6_12mts',
 'num_times_60p_dpd',
 'num_std_12mts',
 'num_sub',
 'num_sub_6mts',
 'num_sub_12mts',
 'num_dbt',
 'num_dbt_12mts',
 'num_lss',
 'recent_level_of_deliq',
 'CC_enq_L12m',
 'PL_enq_L12m',
 'time_since_recent_enq',
 'enq_L3m',
 'NETMONTHLYINCOME',
 'Time_With_Curr_Empr',
 'CC_Flag',
 'PL_Flag',
 'pct_PL_enq_L6m_of_ever',
 'pct_CC_enq_L6m_of_ever',
 'HL_Flag',
 'GL_Flag']

### feature selection is done for cat and num features

## listing all the final features

In [39]:
# listing all the final features
features = columns_to_be_kept_numerical + ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']
df = df[features + ['Approved_Flag']]

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 43 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            42064 non-null  float64
 1   pct_tl_closed_L6M          42064 non-null  float64
 2   Tot_TL_closed_L12M         42064 non-null  int64  
 3   pct_tl_closed_L12M         42064 non-null  float64
 4   Tot_Missed_Pmnt            42064 non-null  int64  
 5   CC_TL                      42064 non-null  int64  
 6   Home_TL                    42064 non-null  int64  
 7   PL_TL                      42064 non-null  int64  
 8   Secured_TL                 42064 non-null  int64  
 9   Unsecured_TL               42064 non-null  int64  
 10  Other_TL                   42064 non-null  int64  
 11  Age_Oldest_TL              42064 non-null  int64  
 12  Age_Newest_TL              42064 non-null  int64  
 13  time_since_recent_payment  42064 non-null  int

In [41]:
df.shape

(42064, 43)

In [42]:
# Label encoding for the categorical features is needed

###### because system can't understand categorical values

In [43]:
df['MARITALSTATUS'].unique()


array(['Married', 'Single'], dtype=object)

In [44]:
df['EDUCATION'].unique()


array(['12TH', 'GRADUATE', 'SSC', 'POST-GRADUATE', 'UNDER GRADUATE',
       'OTHERS', 'PROFESSIONAL'], dtype=object)

In [45]:
df['GENDER'].unique()

array(['M', 'F'], dtype=object)

In [46]:
df['last_prod_enq2'].unique()

array(['PL', 'ConsumerLoan', 'AL', 'CC', 'others', 'HL'], dtype=object)

In [47]:
df['first_prod_enq2'].unique()

array(['PL', 'ConsumerLoan', 'others', 'AL', 'HL', 'CC'], dtype=object)

In [48]:
#Ordinal Feature -- Education
#ssc-1,
#12th-2,
#Graduate-3,
#Under_Graduate--3,
#Post_Graduate-4,
#OTHERS-1,
#Professional-3

In [49]:
# Others has to be verified by the business end user 


In [50]:
df.loc[df['EDUCATION'] == 'SSC',['EDUCATION']]              = 1
df.loc[df['EDUCATION'] == '12TH',['EDUCATION']]             = 2
df.loc[df['EDUCATION'] == 'GRADUATE',['EDUCATION']]         = 3
df.loc[df['EDUCATION'] == 'UNDER GRADUATE',['EDUCATION']]   = 3
df.loc[df['EDUCATION'] == 'POST-GRADUATE',['EDUCATION']]    = 4
df.loc[df['EDUCATION'] == 'OTHERS',['EDUCATION']]           = 1
df.loc[df['EDUCATION'] == 'PROFESSIONAL',['EDUCATION']]     = 3

In [51]:
df['EDUCATION'].value_counts()
df['EDUCATION'] = df['EDUCATION'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 43 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            42064 non-null  float64
 1   pct_tl_closed_L6M          42064 non-null  float64
 2   Tot_TL_closed_L12M         42064 non-null  int64  
 3   pct_tl_closed_L12M         42064 non-null  float64
 4   Tot_Missed_Pmnt            42064 non-null  int64  
 5   CC_TL                      42064 non-null  int64  
 6   Home_TL                    42064 non-null  int64  
 7   PL_TL                      42064 non-null  int64  
 8   Secured_TL                 42064 non-null  int64  
 9   Unsecured_TL               42064 non-null  int64  
 10  Other_TL                   42064 non-null  int64  
 11  Age_Oldest_TL              42064 non-null  int64  
 12  Age_Newest_TL              42064 non-null  int64  
 13  time_since_recent_payment  42064 non-null  int

In [52]:
df_encoded = pd.get_dummies(df, columns=['MARITALSTATUS','GENDER', 'last_prod_enq2' ,'first_prod_enq2'])


In [53]:
df_encoded.info()
k = df_encoded.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 55 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   pct_tl_open_L6M               42064 non-null  float64
 1   pct_tl_closed_L6M             42064 non-null  float64
 2   Tot_TL_closed_L12M            42064 non-null  int64  
 3   pct_tl_closed_L12M            42064 non-null  float64
 4   Tot_Missed_Pmnt               42064 non-null  int64  
 5   CC_TL                         42064 non-null  int64  
 6   Home_TL                       42064 non-null  int64  
 7   PL_TL                         42064 non-null  int64  
 8   Secured_TL                    42064 non-null  int64  
 9   Unsecured_TL                  42064 non-null  int64  
 10  Other_TL                      42064 non-null  int64  
 11  Age_Oldest_TL                 42064 non-null  int64  
 12  Age_Newest_TL                 42064 non-null  int64  
 13  t

In [54]:
df_encoded.tail()

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
42059            0.333               0.00                   0   
42060            0.000               0.25                   1   
42061            0.500               0.50                   1   
42062            0.000               0.00                   1   
42063            0.000               0.00                   0   

       pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
42059                0.00                0      0        0      0           0   
42060                0.25                0      0        0      0           2   
42061                0.50                0      0        0      0           0   
42062                0.50                0      0        0      0           0   
42063                0.00                0      0        0      0           1   

       Unsecured_TL  ...  last_prod_enq2_ConsumerLoan  last_prod_enq2_HL  \
42059             3  ...                         True              False   
42060             2  ...                        False              False   
42061             2  ...                         True              False   
42062             2  ...                         True              False   
42063             0  ...                        False              False   

       last_prod_enq2_PL  last_prod_enq2_others  first_prod_enq2_AL  \
42059              False                  False               False   
42060              False                   True               False   
42061              False                  False               False   
42062              False                  False               False   
42063              False                   True               False   

       first_prod_enq2_CC  first_prod_enq2_ConsumerLoan  first_prod_enq2_HL  \
42059               False                          True               False   
42060               False                         False               False   
42061               False                         False               False   
42062               False                         False               False   
42063               False                         False               False   

       first_prod_enq2_PL  first_prod_enq2_others  
42059               False                   False  
42060               False                    True  
42061               False                    True  
42062               False                    True  
42063               False                    True  

[5 rows x 55 columns]

In [55]:
k

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
count     42064.000000       42064.000000        42064.000000   
mean          0.179032           0.097783            0.825504   
std           0.278043           0.210957            1.537208   
min           0.000000           0.000000            0.000000   
25%           0.000000           0.000000            0.000000   
50%           0.000000           0.000000            0.000000   
75%           0.333000           0.100000            1.000000   
max           1.000000           1.000000           33.000000   

       pct_tl_closed_L12M  Tot_Missed_Pmnt         CC_TL       Home_TL  \
count        42064.000000     42064.000000  42064.000000  42064.000000   
mean             0.160365         0.525746      0.145921      0.076241   
std              0.258831         1.106442      0.549314      0.358582   
min              0.000000         0.000000      0.000000      0.000000   
25%              0.000000         0.000000      0.000000      0.000000   
50%              0.000000         0.000000      0.000000      0.000000   
75%              0.250000         1.000000      0.000000      0.000000   
max              1.000000        34.000000     27.000000     10.000000   

              PL_TL    Secured_TL  Unsecured_TL  ...       enq_L3m  \
count  42064.000000  42064.000000  42064.000000  ...  42064.000000   
mean       0.328000      2.921334      2.341646  ...      1.230458   
std        0.916368      6.379764      3.405397  ...      2.069461   
min        0.000000      0.000000      0.000000  ...      0.000000   
25%        0.000000      0.000000      0.000000  ...      0.000000   
50%        0.000000      1.000000      1.000000  ...      1.000000   
75%        0.000000      3.000000      3.000000  ...      2.000000   
max       29.000000    235.000000     55.000000  ...     42.000000   

       NETMONTHLYINCOME  Time_With_Curr_Empr       CC_Flag       PL_Flag  \
count      4.206400e+04         42064.000000  42064.000000  42064.000000   
mean       2.692990e+04           110.345783      0.102962      0.193063   
std        2.084300e+04            75.629967      0.303913      0.394707   
min        0.000000e+00             0.000000      0.000000      0.000000   
25%        1.800000e+04            61.000000      0.000000      0.000000   
50%        2.400000e+04            92.000000      0.000000      0.000000   
75%        3.100000e+04           131.000000      0.000000      0.000000   
max        2.500000e+06          1020.000000      1.000000      1.000000   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever       HL_Flag  \
count            42064.000000            42064.000000  42064.000000   
mean                 0.195497                0.064186      0.252235   
std                  0.367414                0.225989      0.434300   
min                  0.000000                0.000000      0.000000   
25%                  0.000000                0.000000      0.000000   
50%                  0.000000                0.000000      0.000000   
75%                  0.000000                0.000000      1.000000   
max                  1.000000                1.000000      1.000000   

            GL_Flag     EDUCATION  
count  42064.000000  42064.000000  
mean       0.056580      2.313689  
std        0.231042      0.871070  
min        0.000000      1.000000  
25%        0.000000      2.000000  
50%        0.000000      2.000000  
75%        0.000000      3.000000  
max        1.000000      4.000000  

[8 rows x 38 columns]

In [56]:
df_encoded.columns

Index(['pct_tl_open_L6M', 'pct_tl_closed_L6M', 'Tot_TL_closed_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL', 'Home_TL', 'PL_TL',
       'Secured_TL', 'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL',
       'Age_Newest_TL', 'time_since_recent_payment',
       'max_recent_level_of_deliq', 'num_deliq_6_12mts', 'num_times_60p_dpd',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_12mts', 'num_lss', 'recent_level_of_deliq', 'CC_enq_L12m',
       'PL_enq_L12m', 'time_since_recent_enq', 'enq_L3m', 'NETMONTHLYINCOME',
       'Time_With_Curr_Empr', 'CC_Flag', 'PL_Flag', 'pct_PL_enq_L6m_of_ever',
       'pct_CC_enq_L6m_of_ever', 'HL_Flag', 'GL_Flag', 'EDUCATION',
       'Approved_Flag', 'MARITALSTATUS_Married', 'MARITALSTATUS_Single',
       'GENDER_F', 'GENDER_M', 'last_prod_enq2_AL', 'last_prod_enq2_CC',
       'last_prod_enq2_ConsumerLoan', 'last_prod_enq2_HL', 'last_prod_enq2_PL',
       'last_prod_enq2_others', 'first_prod_enq2_AL',

In [57]:
df_encoded.describe().T

count          mean           std  min      25%  \
pct_tl_open_L6M            42064.0      0.179032      0.278043  0.0      0.0   
pct_tl_closed_L6M          42064.0      0.097783      0.210957  0.0      0.0   
Tot_TL_closed_L12M         42064.0      0.825504      1.537208  0.0      0.0   
pct_tl_closed_L12M         42064.0      0.160365      0.258831  0.0      0.0   
Tot_Missed_Pmnt            42064.0      0.525746      1.106442  0.0      0.0   
CC_TL                      42064.0      0.145921      0.549314  0.0      0.0   
Home_TL                    42064.0      0.076241      0.358582  0.0      0.0   
PL_TL                      42064.0      0.328000      0.916368  0.0      0.0   
Secured_TL                 42064.0      2.921334      6.379764  0.0      0.0   
Unsecured_TL               42064.0      2.341646      3.405397  0.0      0.0   
Other_TL                   42064.0      1.116489      2.486801  0.0      0.0   
Age_Oldest_TL              42064.0     46.498074     42.109230  0.0     14.0   
Age_Newest_TL              42064.0     13.970046     18.835191  0.0      4.0   
time_since_recent_payment  42064.0    218.601607    422.282417  2.0     51.0   
max_recent_level_of_deliq  42064.0     14.314758     54.056303  0.0      0.0   
num_deliq_6_12mts          42064.0      0.336963      1.097356  0.0      0.0   
num_times_60p_dpd          42064.0      0.438879      2.148400  0.0      0.0   
num_std_12mts              42064.0      3.279978      7.566312  0.0      0.0   
num_sub                    42064.0      0.063831      0.799989  0.0      0.0   
num_sub_6mts               42064.0      0.002211      0.081704  0.0      0.0   
num_sub_12mts              42064.0      0.009224      0.220786  0.0      0.0   
num_dbt                    42064.0      0.024510      0.621890  0.0      0.0   
num_dbt_12mts              42064.0      0.004279      0.184461  0.0      0.0   
num_lss                    42064.0      0.016713      0.573762  0.0      0.0   
recent_level_of_deliq      42064.0     11.803918     46.422091  0.0      0.0   
CC_enq_L12m                42064.0      0.268924      1.019459  0.0      0.0   
PL_enq_L12m                42064.0      0.779194      1.802092  0.0      0.0   
time_since_recent_enq      42064.0    264.854507    466.585002  0.0      9.0   
enq_L3m                    42064.0      1.230458      2.069461  0.0      0.0   
NETMONTHLYINCOME           42064.0  26929.896800  20842.999867  0.0  18000.0   
Time_With_Curr_Empr        42064.0    110.345783     75.629967  0.0     61.0   
CC_Flag                    42064.0      0.102962      0.303913  0.0      0.0   
PL_Flag                    42064.0      0.193063      0.394707  0.0      0.0   
pct_PL_enq_L6m_of_ever     42064.0      0.195497      0.367414  0.0      0.0   
pct_CC_enq_L6m_of_ever     42064.0      0.064186      0.225989  0.0      0.0   
HL_Flag                    42064.0      0.252235      0.434300  0.0      0.0   
GL_Flag                    42064.0      0.056580      0.231042  0.0      0.0   
EDUCATION                  42064.0      2.313689      0.871070  1.0      2.0   

                               50%        75%        max  
pct_tl_open_L6M                0.0      0.333        1.0  
pct_tl_closed_L6M              0.0      0.100        1.0  
Tot_TL_closed_L12M             0.0      1.000       33.0  
pct_tl_closed_L12M             0.0      0.250        1.0  
Tot_Missed_Pmnt                0.0      1.000       34.0  
CC_TL                          0.0      0.000       27.0  
Home_TL                        0.0      0.000       10.0  
PL_TL                          0.0      0.000       29.0  
Secured_TL                     1.0      3.000      235.0  
Unsecured_TL                   1.0      3.000       55.0  
Other_TL                       0.0      1.000       80.0  
Age_Oldest_TL                 34.0     65.000      385.0  
Age_Newest_TL                  7.0     16.000      359.0  
time_since_recent_payment     71.0    146.000     6065.0  
max_recent_level_of_del

In [58]:
X = df_encoded. drop ( ['Approved_Flag'], axis = 1 )

In [ ]:
X

## 1.Random Forest

In [59]:
# Machine Learing model fitting
# Data processing



y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )


In [60]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [61]:
rf_classifier = RandomForestClassifier(n_estimators = 200, random_state=42)

In [62]:
rf_classifier.fit(x_train, y_train)

RandomForestClassifier(n_estimators=200, random_state=42)

In [63]:
y_pred = rf_classifier.predict(x_test)

In [64]:
accuracy = accuracy_score(y_test, y_pred)
print ()
print(f'Accuracy: {accuracy}')
print ()
precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)



Accuracy: 0.7636990372043266



In [65]:
for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()

Class p1:
Precision: 0.8370457209847597
Recall: 0.7041420118343196
F1 Score: 0.7648634172469203

Class p2:
Precision: 0.7957519116397621
Recall: 0.9282457879088206
F1 Score: 0.8569075937785909

Class p3:
Precision: 0.4423380726698262
Recall: 0.21132075471698114
F1 Score: 0.28600612870275793

Class p4:
Precision: 0.7178502879078695
Recall: 0.7269193391642371
F1 Score: 0.7223563495895703



## 2.Xgboost

In [66]:
import xgboost as xgb
from sklearn.preprocessing import LabelEncoder

In [67]:
xgb_classifier = xgb.XGBClassifier(objective='multi:softmax',  num_class=4)

In [68]:
xgb_classifier

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None, num_class=4,
              num_parallel_tree=None, ...)

In [69]:
y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )


In [70]:
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)


In [71]:
x_train, x_test, y_train, y_test = train_test_split(x, y_encoded, test_size=0.2, random_state=42)

In [72]:
xgb_classifier.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None, num_class=4,
              num_parallel_tree=None, ...)

In [73]:
y_pred = xgb_classifier.predict(x_test)

In [74]:
accuracy = accuracy_score(y_test, y_pred)
print ()
print(f'Accuracy: {accuracy:.4f}')
print ()



Accuracy: 0.7783



In [75]:
precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)


In [76]:
for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()

Class p1:
Precision: 0.823906083244397
Recall: 0.7613412228796844
F1 Score: 0.7913890312660173

Class p2:
Precision: 0.8255418233924413
Recall: 0.913577799801784
F1 Score: 0.8673315769665036

Class p3:
Precision: 0.4756380510440835
Recall: 0.30943396226415093
F1 Score: 0.3749428440786465

Class p4:
Precision: 0.7342386032977691
Recall: 0.7356656948493683
F1 Score: 0.7349514563106796



## 3.Decision Tree

In [77]:
from sklearn.tree import DecisionTreeClassifier

In [78]:
y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )

In [79]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)


In [80]:
dt_model = DecisionTreeClassifier(max_depth=20, min_samples_split=10)
dt_model.fit(x_train, y_train)


DecisionTreeClassifier(max_depth=20, min_samples_split=10)

In [81]:
y_pred = dt_model.predict(x_test)

In [82]:
accuracy = accuracy_score(y_test, y_pred)
print ()
print(f"Accuracy: {accuracy:.4f}")
print ()



Accuracy: 0.7108



In [83]:
precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)

In [84]:
for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()

Class p1:
Precision: 0.7250996015936255
Recall: 0.717948717948718
F1 Score: 0.7215064420218038

Class p2:
Precision: 0.8085725368502715
Recall: 0.8263627353815659
F1 Score: 0.8173708459954906

Class p3:
Precision: 0.3455706304868316
Recall: 0.3267924528301887
F1 Score: 0.3359193173002328

Class p4:
Precision: 0.65
Recall: 0.6316812439261419
F1 Score: 0.6407097092163628



### Xgboost is giving best results

#### so we will pick it and further finetune it

## Hyperparameter tuning for xgboost 

In [85]:
# Define the hyperparameter grid
param_grid = {
   'colsample_bytree': [0.1, 0.3, 0.5, 0.7, 0.9],
   'learning_rate'   : [0.001, 0.01, 0.1, 1],
   'max_depth'       : [3, 5, 8, 10],
   'alpha'           : [1, 10, 100],
   'n_estimators'    : [10,50,100]
}

In [86]:
index = 0

In [87]:
answers_grid = {
     'combination'       :[],
     'train_Accuracy'    :[],
     'test_Accuracy'     :[],
     'colsample_bytree'  :[],
     'learning_rate'     :[],
     'max_depth'         :[],
     'alpha'             :[],
     'n_estimators'      :[]

     }


In [88]:
# Loop through each combination of hyperparameters
for colsample_bytree in param_grid['colsample_bytree']:
    for learning_rate in param_grid['learning_rate']:
        for max_depth in param_grid['max_depth']:
            for alpha in param_grid['alpha']:
                for n_estimators in param_grid['n_estimators']:
                    print("colsample_bytree:", colsample_bytree)
                    print("learning_rate:", learning_rate)
                    print("max_depth:", max_depth)
                    print("alpha:", alpha)
                    print("n_estimators:", n_estimators)
                    print()  # Add a blank line for readability
                    
                    index = index + 1
             

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 1
n_estimators: 10

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 1
n_estimators: 50

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 1
n_estimators: 100

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 10
n_estimators: 10

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 10
n_estimators: 50

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 10
n_estimators: 100

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 100
n_estimators: 10

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 100
n_estimators: 50

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 3
alpha: 100
n_estimators: 100

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 5
alpha: 1
n_estimators: 10

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 5
alpha: 1
n_estimators: 50

colsample_bytree: 0.1
learning_rate: 0.001
max_depth: 5
alpha: 1
n_estimato

In [89]:
 # Define and train the XGBoost model
model = xgb.XGBClassifier(objective='multi:softmax',  
                                        num_class=4,
                                        colsample_bytree = colsample_bytree,
                                        learning_rate = learning_rate,
                                        max_depth = max_depth,
                                        alpha = alpha,
                                        n_estimators = n_estimators)
            

In [90]:
y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )

In [91]:
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

In [92]:
x_train, x_test, y_train, y_test = train_test_split(x, y_encoded, test_size=0.2, random_state=42)

In [93]:
model.fit(x_train, y_train)

XGBClassifier(alpha=100, base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.9, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=10, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=100, n_jobs=None, num_class=4, ...)

In [94]:
 # Predict on training and testing sets
y_pred_train = model.predict(x_train)
y_pred_test = model.predict(x_test)


In [95]:
y_pred_train

array([0, 0, 1, ..., 1, 3, 0])

In [96]:
y_pred_test

array([1, 3, 1, ..., 2, 2, 3])

In [97]:
# Calculate train and test results
train_accuracy =  accuracy_score (y_train, y_pred_train)
test_accuracy  =  accuracy_score (y_test , y_pred_test)

In [98]:
print('train_Accuarcy:',train_accuracy)

train_Accuarcy: 0.786247065466108


In [99]:
print('test_Accuarcy:',test_accuracy)

test_Accuarcy: 0.7752288125520028


### No overfitting or underfitting both train and test values giving same accuracy

##### we proceed to further

In [100]:
 # Include into the lists
answers_grid ['combination']   .append(index)
answers_grid ['train_Accuracy']    .append(train_accuracy)
answers_grid ['test_Accuracy']     .append(test_accuracy)
answers_grid ['colsample_bytree']   .append(colsample_bytree)
answers_grid ['learning_rate']      .append(learning_rate)
answers_grid ['max_depth']          .append(max_depth)
answers_grid ['alpha']              .append(alpha)
answers_grid ['n_estimators']       .append(n_estimators)

In [101]:
answers_grid # best parameters 

{'combination': [720],
 'train_Accuracy': [0.786247065466108],
 'test_Accuracy': [0.7752288125520028],
 'colsample_bytree': [0.9],
 'learning_rate': [1],
 'max_depth': [10],
 'alpha': [100],
 'n_estimators': [100]}

In [102]:
# Print results for this combination
print(f"Combination {index}")
print(f"colsample_bytree: {colsample_bytree}, learning_rate: {learning_rate}, max_depth: {max_depth}, alpha: {alpha}, n_estimators: {n_estimators}")


Combination 720
colsample_bytree: 0.9, learning_rate: 1, max_depth: 10, alpha: 100, n_estimators: 100


In [103]:
print(f"Train Accuracy: {train_accuracy:.2f}")
print(f"Test Accuracy : {test_accuracy :.2f}")
print("-" * 30)

Train Accuracy: 0.79
Test Accuracy : 0.78
------------------------------


In [104]:
#model.predict([[]])

#### Predict for unseen dataset

In [105]:
a3 = pd.read_excel('unseen_Dataset.xlsx')

In [106]:
col_in_df = list(df.columns)
col_in_df.pop(42)

'Approved_Flag'

In [107]:
df_unseen = a3[col_in_df]

In [108]:
df_unseen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 42 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            100 non-null    float64
 1   pct_tl_closed_L6M          100 non-null    float64
 2   Tot_TL_closed_L12M         100 non-null    int64  
 3   pct_tl_closed_L12M         100 non-null    float64
 4   Tot_Missed_Pmnt            100 non-null    int64  
 5   CC_TL                      100 non-null    int64  
 6   Home_TL                    100 non-null    int64  
 7   PL_TL                      100 non-null    int64  
 8   Secured_TL                 100 non-null    int64  
 9   Unsecured_TL               100 non-null    int64  
 10  Other_TL                   100 non-null    int64  
 11  Age_Oldest_TL              100 non-null    int64  
 12  Age_Newest_TL              100 non-null    int64  
 13  time_since_recent_payment  100 non-null    int64  


In [109]:
df_unseen['MARITALSTATUS'].unique()

array(['Married', 'Single'], dtype=object)

In [110]:
df_unseen['EDUCATION'].unique()

array(['12TH', 'GRADUATE', 'SSC', 'POST-GRADUATE', 'UNDER GRADUATE',
       'OTHERS'], dtype=object)

In [111]:
df_unseen['GENDER'].unique()

array(['M', 'F'], dtype=object)

In [112]:
df_unseen['last_prod_enq2'] 

0               PL
1     ConsumerLoan
2     ConsumerLoan
3               AL
4     ConsumerLoan
          ...     
95    ConsumerLoan
96    ConsumerLoan
97          others
98          others
99          others
Name: last_prod_enq2, Length: 100, dtype: object

In [113]:
df_unseen['first_prod_enq2'].unique()

array(['PL', 'ConsumerLoan', 'others', 'AL', 'HL', 'CC'], dtype=object)

### Creating Accuracy excel sheet of 720 Combinations

In [114]:
import pandas as pd

# Assuming answers_grid is a dictionary with lists as values
answers_grid = {
    'combination': [],
    'train_Accuracy': [],
    'test_Accuracy': [],
    'colsample_bytree': [],
    'learning_rate': [],
    'max_depth': [],
    'alpha': [],
    'n_estimators': []
}

# Loop through each combination of hyperparameters
index = 1
for colsample_bytree in param_grid['colsample_bytree']:
    for learning_rate in param_grid['learning_rate']:
        for max_depth in param_grid['max_depth']:
            for alpha in param_grid['alpha']:
                for n_estimators in param_grid['n_estimators']:
                   
                    train_accuracy =  accuracy_score (y_train, y_pred_train)
                    test_accuracy  =  accuracy_score (y_test , y_pred_test)
                    
                    answers_grid['combination'].append(index)
                    answers_grid['train_Accuracy'].append(train_accuracy)
                    answers_grid['test_Accuracy'].append(test_accuracy)
                    answers_grid['colsample_bytree'].append(colsample_bytree)
                    answers_grid['learning_rate'].append(learning_rate)
                    answers_grid['max_depth'].append(max_depth)
                    answers_grid['alpha'].append(alpha)
                    answers_grid['n_estimators'].append(n_estimators)
                    
                    index += 1
# Convert answers_grid to a DataFrame
data = pd.DataFrame(answers_grid)

# Specify the file path where you want to save the Excel file
excel_file_path = "Accuracy_score.xlsx"

# Write the DataFrame to Excel
data.to_excel(excel_file_path, index=False)
 
print("Excel file created successfully!")



Excel file created successfully!


In [115]:
df_unseen.loc[df_unseen['EDUCATION'] == 'SSC',['EDUCATION']]             = 1
df_unseen.loc[df_unseen['EDUCATION'] == '12TH',['EDUCATION']]            = 2
df_unseen.loc[df_unseen['EDUCATION'] == 'GRADUATE',['EDUCATION']]        = 3
df_unseen.loc[df_unseen['EDUCATION'] == 'UNDER GRADUATE',['EDUCATION']]  = 3
df_unseen.loc[df_unseen['EDUCATION'] == 'POST-GRADUATE',['EDUCATION']]   = 4
df_unseen.loc[df_unseen['EDUCATION'] == 'OTHERS',['EDUCATION']]          = 1
df_unseen.loc[df_unseen['EDUCATION'] == 'PROFESSIONAL',['EDUCATION']]    = 3

In [116]:
df_unseen['EDUCATION'].value_counts()
df_unseen['EDUCATION'] = df_unseen['EDUCATION'].astype(int)
df_unseen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 42 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            100 non-null    float64
 1   pct_tl_closed_L6M          100 non-null    float64
 2   Tot_TL_closed_L12M         100 non-null    int64  
 3   pct_tl_closed_L12M         100 non-null    float64
 4   Tot_Missed_Pmnt            100 non-null    int64  
 5   CC_TL                      100 non-null    int64  
 6   Home_TL                    100 non-null    int64  
 7   PL_TL                      100 non-null    int64  
 8   Secured_TL                 100 non-null    int64  
 9   Unsecured_TL               100 non-null    int64  
 10  Other_TL                   100 non-null    int64  
 11  Age_Oldest_TL              100 non-null    int64  
 12  Age_Newest_TL              100 non-null    int64  
 13  time_since_recent_payment  100 non-null    int64  


In [117]:
df_encoded_unseen = pd.get_dummies(df_unseen, columns=['MARITALSTATUS','GENDER', 'last_prod_enq2' ,'first_prod_enq2'])


In [118]:
df_encoded_unseen

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
0             0.000                0.0                   0   
1             0.000                0.0                   0   
2             0.125                0.0                   0   
3             0.000                0.0                   0   
4             0.000                0.0                   1   
..              ...                ...                 ...   
95            0.000                0.0                   0   
96            0.222                0.0                   2   
97            0.000                0.0                   0   
98            0.000                1.0                   1   
99            0.500                0.0                   0   

    pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
0                0.000                0      0        0      4           1   
1                0.000                0      0        0      0           0   
2                0.000                1      0        0      0           2   
3                0.000                0      0        0      0           3   
4                0.167                0      0        0      0           6   
..                 ...              ...    ...      ...    ...         ...   
95               0.000                0      0        0      1           1   
96               0.222                0      0        0      0           1   
97               0.000                0      0        0      0           1   
98               1.000                0      0        0      0           1   
99               0.000                0      0        0      0           1   

    Unsecured_TL  ...  last_prod_enq2_ConsumerLoan  last_prod_enq2_HL  \
0              4  ...                        False              False   
1              1  ...                         True              False   
2              6  ...                         True              False   
3              0  ...                        False              False   
4              0  ...                         True              False   
..           ...  ...                          ...                ...   
95             5  ...                         True              False   
96             8  ...                         True              False   
97             0  ...                        False              False   
98             0  ...                        False              False   
99             1  ...                        False              False   

    last_prod_enq2_PL  last_prod_enq2_others  first_prod_enq2_AL  \
0                True                  False               False   
1               False                  False               False   
2               False                  False               False   
3               False                  False                True   
4               False                  False               False   
..                ...                    ...                 ...   
95              False                  False               False   
96              False                  False               False   
97              False                   True               False   
98              False                   True               False   
99              False                   True               False   

    first_prod_enq2_CC  first_prod_enq2_ConsumerLoan  first_prod_enq2_HL  \
0                False                         False               False   
1                False                          True               False   
2                False                         False               False   
3                False                         False               False   
4                False                         False               False   
..                 ...                           ...                 ...   
95               False                         False               False   
96               Fal

In [119]:
df_encoded_unseen.info()
k = df_encoded_unseen.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 54 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   pct_tl_open_L6M               100 non-null    float64
 1   pct_tl_closed_L6M             100 non-null    float64
 2   Tot_TL_closed_L12M            100 non-null    int64  
 3   pct_tl_closed_L12M            100 non-null    float64
 4   Tot_Missed_Pmnt               100 non-null    int64  
 5   CC_TL                         100 non-null    int64  
 6   Home_TL                       100 non-null    int64  
 7   PL_TL                         100 non-null    int64  
 8   Secured_TL                    100 non-null    int64  
 9   Unsecured_TL                  100 non-null    int64  
 10  Other_TL                      100 non-null    int64  
 11  Age_Oldest_TL                 100 non-null    int64  
 12  Age_Newest_TL                 100 non-null    int64  
 13  time_s

In [120]:
k

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
count       100.000000         100.000000           100.00000   
mean          0.191230           0.116210             0.71000   
std           0.294655           0.253891             1.13079   
min           0.000000           0.000000             0.00000   
25%           0.000000           0.000000             0.00000   
50%           0.000000           0.000000             0.00000   
75%           0.343500           0.091000             1.00000   
max           1.000000           1.000000             5.00000   

       pct_tl_closed_L12M  Tot_Missed_Pmnt       CC_TL     Home_TL  \
count           100.00000       100.000000  100.000000  100.000000   
mean              0.16822         0.410000    0.090000    0.040000   
std               0.27274         0.792579    0.320826    0.242878   
min               0.00000         0.000000    0.000000    0.000000   
25%               0.00000         0.000000    0.000000    0.000000   
50%               0.00000         0.000000    0.000000    0.000000   
75%               0.25000         0.250000    0.000000    0.000000   
max               1.00000         3.000000    2.000000    2.000000   

            PL_TL  Secured_TL  Unsecured_TL  ...    enq_L3m  NETMONTHLYINCOME  \
count  100.000000  100.000000    100.000000  ...  100.00000        100.000000   
mean     0.410000    2.330000      2.300000  ...    1.09000      25764.300000   
std      1.005992    4.811718      2.544553  ...    1.83179      13707.266015   
min      0.000000    0.000000      0.000000  ...    0.00000          0.000000   
25%      0.000000    0.000000      0.000000  ...    0.00000      15875.000000   
50%      0.000000    1.000000      1.500000  ...    0.00000      24000.000000   
75%      0.000000    2.000000      3.000000  ...    1.00000      30000.000000   
max      7.000000   35.000000     10.000000  ...   10.00000      65000.000000   

       Time_With_Curr_Empr    CC_Flag     PL_Flag  pct_PL_enq_L6m_of_ever  \
count           100.000000  100.00000  100.000000              100.000000   
mean            108.330000    0.08000    0.230000                0.154610   
std              65.996029    0.27266    0.422953                0.335204   
min              13.000000    0.00000    0.000000                0.000000   
25%              60.750000    0.00000    0.000000                0.000000   
50%             101.500000    0.00000    0.000000                0.000000   
75%             130.000000    0.00000    0.000000                0.000000   
max             306.000000    1.00000    1.000000                1.000000   

       pct_CC_enq_L6m_of_ever     HL_Flag     GL_Flag   EDUCATION  
count              100.000000  100.000000  100.000000  100.000000  
mean                 0.093330    0.160000    0.030000    2.250000  
std                  0.279607    0.368453    0.171447    0.903137  
min                  0.000000    0.000000    0.000000    1.000000  
25%                  0.000000    0.000000    0.000000    1.000000  
50%                  0.000000    0.000000    0.000000    2.000000  
75%                  0.000000    0.000000    0.000000    3.000000  
max                  1.000000    1.000000    1.000000    4.000000  

[8 rows x 38 columns]

In [121]:
model = xgb.XGBClassifier(objective='multi:softmax',
                                num_classes=4,
                                colsample_bytree=0.9,
                                learning_rate = 1,
                                max_depth = 3,
                                alpha = 10,
                                n_estimators = 100)

In [122]:
model.fit(x_train,y_train)

XGBClassifier(alpha=10, base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.9, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=100, n_jobs=None, num_classes=4, ...)

In [123]:
y_pred_unseen = model.predict(df_encoded_unseen)

In [124]:
a3['Target_variable'] = y_pred_unseen

In [125]:
y_pred_unseen

array([1, 1, 1, 0, 1, 2, 2, 1, 1, 1, 1, 3, 1, 0, 2, 1, 1, 0, 3, 1, 0, 1,
       1, 3, 1, 2, 1, 1, 2, 1, 3, 1, 2, 1, 1, 3, 1, 3, 1, 1, 1, 2, 1, 2,
       3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 2, 0,
       1, 1, 1, 1, 1, 3, 3, 1, 1, 1, 1, 1])

In [126]:
df['Approved_Flag'].value_counts()

Approved_Flag
P2    25452
P3     6440
P4     5264
P1     4908
Name: count, dtype: int64

In [127]:
data = {'Target_variable': y_pred_unseen}
dataframe = pd.DataFrame(data)

# Specify the file path where you want to save the Excel file
file_path = 'predicted_values.xlsx'

# Export DataFrame to Excel
dataframe.to_excel(file_path, index=False)

print(f"Predicted values saved to '{file_path}'.")

Predicted values saved to 'predicted_values.xlsx'.


In [128]:
predicted_values = pd.read_excel('predicted_values.xlsx')

In [129]:
predicted_values.value_counts() 

Target_variable
1                  71
3                  13
2                   9
0                   7
Name: count, dtype: int64

In [130]:
#Here 1 = p1, 2 = p2, 3 = p3, 4 = p4  

In [131]:
#print runtime 

In [132]:
end_time = time.time()

In [133]:
elapsed_time = end_time - start_time

In [134]:
print(f"total run time of the program."+ str(round(elapsed_time,2)) + 'sec')

total run time of the program.152.06sec


In [135]:
import pickle 

In [137]:
# Load the model from the pickle file
with open(pickle_path, 'rb') as file:
    loaded_model = pickle.load(file)